clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -459,7 +467,506 @@ void SemaHLSL::handleResourceClassAttr(Decl *D, const ParsedAttr &AL) {
   D->addAttr(HLSLResourceClassAttr::Create(getASTContext(), RC, ArgLoc));
 }
 
-void SemaHLSL::handleResourceBindingAttr(Decl *D, const ParsedAttr &AL) {
+struct RegisterBindingFlags {
+  bool Resource = false;
+  bool UDT = false;
+  bool Other = false;
+  bool Basic = false;
+
+  bool SRV = false;
+  bool UAV = false;
+  bool CBV = false;
+  bool Sampler = false;
+
+  bool ContainsNumeric = false;
+  bool DefaultGlobals = false;
+};
+
+bool isDeclaredWithinCOrTBuffer(const Decl *TheDecl) {
+  if (!TheDecl)
+    return false;
+
+  // Traverse up the parent contexts
+  const DeclContext *context = TheDecl->getDeclContext();
+  if (isa<HLSLBufferDecl>(context)) {
+    return true;
+  }
+
+  return false;
+}
+
+const CXXRecordDecl *getRecordDeclFromVarDecl(VarDecl *VD) {
+  const Type *Ty = VD->getType()->getPointeeOrArrayElementType();
+  assert(Ty && "Resource class must have an element type.");
+
+  if (const auto *TheBuiltinTy = dyn_cast<BuiltinType>(Ty))
+    return nullptr;
+
+  const CXXRecordDecl *TheRecordDecl = Ty->getAsCXXRecordDecl();
+  assert(TheRecordDecl &&
+         "Resource class should have a resource type declaration.");
+
+  if (auto TDecl = dyn_cast<ClassTemplateSpecializationDecl>(TheRecordDecl))
+    TheRecordDecl = TDecl->getSpecializedTemplate()->getTemplatedDecl();
+  TheRecordDecl = TheRecordDecl->getCanonicalDecl();
+  return TheRecordDecl;
+}
+
+const HLSLResourceClassAttr *
+getHLSLResourceClassAttrFromEitherDecl(VarDecl *VD,
+                                       HLSLBufferDecl *CBufferOrTBuffer) {
+
```

Comment:
Near the top of `DiagnoseHLSLRegisterAttribute`, there's an assert: 
`assert(((TheVarDecl && !CBufferOrTBuffer) || (!TheVarDecl && CBufferOrTBuffer)) && "either TheVarDecl or BufferOrTBuffer should be set");`
It will trigger if both are set. If both are set, there's been some sort of bug or error, because it shouldn't be possible for the decl (called `TheDecl` in that function) to be castable to both a vardecl and an hlslbufferdecl. We expect that `cbuffer` and `tbuffer` are represented with an `HLSLBufferDecl`, and that everything else (e.g, RWBuffer) is represented with a `VarDecl`, and neither are subtypes of each other.

Reply:
I'm not convinced that combining this functionality in a single function is all the helpful.  I see this function called in two places:

1. Line 701, in HLSLFillRegisterBindingFlags.  The context where it is called the following is true: `!CBufferOrTbuffer && TheVarDecl`. So in this case it could directly call a `getHLSLResourceClassAttr(VarDecl*)` function.
2. Line 850, in DiagnoseHLSLRegisterAttribute. In this context, I don't think it would be too heinous to have to figure out which one to call, which I think would make the number 1 case above a bit clearer.

Reply:
I've cleaned up the function so that it only handles the VarDecl case, and parameterized it so it can take the attribute it's supposed to look for as well.